sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,263 @@
+        {
+            /// <summary>The period of inactivity after which an item would be removed from the pool</summary>
+            public static readonly TimeSpan DefaultRemoveAfterDuration = TimeSpan.FromMinutes(10);
+
+            /// <summary>
+            ///   A <see cref="TransportProducer"/> assigned to a partition.
+            /// </summary>
+            ///
+            public TransportProducer Producer { get; }
+
+            /// <summary>
+            ///   A set of unique identifiers used to track which instances of a <see cref="PoolItem"/> are active.
+            /// </summary>
+            ///
+            public ConcurrentDictionary<string, byte> ActiveInstances { get; } = new ConcurrentDictionary<string, byte>();
+
+            /// <summary>
+            ///   The UTC date and time when a <see cref="PoolItem"/> will become eligible for eviction.
+            /// </summary>
+            ///
+            public DateTimeOffset RemoveAfter { get; set; }
+
+            /// <summary>
+            ///   Initializes a new instance of the <see cref="PoolItem"/> class with a default timespan of one minute.
+            /// </summary>
+            ///
+            /// <param name="producer">An Event Hub transport-specific producer specific to a given partition</param>
+            /// <param name="removeAfter">The UTC date and time when a <see cref="PoolItem"/> will become eligible for eviction.</param>
+            ///
+            public PoolItem(TransportProducer producer, DateTimeOffset? removeAfter = default)
+            {
+                Argument.AssertNotNull(producer, nameof(producer));
+
+                Producer = producer;
+                RemoveAfter = removeAfter ?? DateTimeOffset.UtcNow.Add(TimeSpan.FromMinutes(1));
+            }
+        }
+
+        /// <summary>
+        ///   A class wrapping a <see cref="TransportProducer"/>, triggering a clean-up when the object
+        ///   is disposed.
+        /// </summary>
+        ///
+        internal class PooledProducer: IAsyncDisposable
+        {
+            /// <summary>
+            ///   A function responsible of cleaning up the resources in use.
+            /// </summary>
+            ///
+            private Func<TransportProducer, Task> CleanUp { get; }
+
+            /// <summary>
+            ///   A <see cref="TransportProducer"/> assigned to a partition.
+            /// </summary>
+            ///
+            public TransportProducer Item { get; }
+
+            /// <summary>
+            ///   Initializes a new instance of the <see cref="PooledProducer"/> class.
+            /// </summary>
+            ///
+            /// <param name="producer">An Event Hub transport-specific producer specific to a given partition</param>
+            /// <param name="cleanUp">The function responsible of cleaning up the resources in use.</param>
+            ///
+            public PooledProducer(TransportProducer producer, Func<TransportProducer, Task> cleanUp)
```

Comment:
(extreme) nit:  Would you be so kind as to split the params into separate lines?